src/mpid/ch4/netmod/ucx/ucx_am.h

```diff
@@ -11,52 +11,85 @@
 
 #include "ucx_impl.h"
 
+#define MPIDI_UCX_SHORT_DATA_SZ  16 * 1024
+
 MPL_STATIC_INLINE_PREFIX void MPIDI_UCX_am_isend_callback(void *request, ucs_status_t status)
 {
     MPIDI_UCX_ucp_request_t *ucp_request = (MPIDI_UCX_ucp_request_t *) request;
-    MPIR_Request *req = ucp_request->req;
-    int handler_id = req->dev.ch4.am.netmod_am.ucx.handler_id;
 
     MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDI_UCX_AM_ISEND_CALLBACK);
     MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDI_UCX_AM_ISEND_CALLBACK);
 
-    MPL_free(req->dev.ch4.am.netmod_am.ucx.pack_buffer);
-    req->dev.ch4.am.netmod_am.ucx.pack_buffer = NULL;
-    MPIDIG_global.origin_cbs[handler_id] (req);
-    ucp_request->req = NULL;
+    if (ucp_request->u.am_send.pack_buffer) {
+        MPL_free(ucp_request->u.am_send.pack_buffer);
+    }
+    if (ucp_request->u.am_send.sreq) {
```

Comment:
If it coming from a `MPIDI_NM_am_send_hdr`, we won't have a request.

Reply:
Also I am paving the way for converging the `MPIDI_NM_am_xxx` APIs at some point.

Reply:
> Also I am paving the way for converging the `MPIDI_NM_am_xxx` APIs at some point.

Why?

Reply:
am_reply is used only to use contexid instead of comm. But in reality we do contextid to comm conversion anyway. Besides, comm could be skipped once we always use global ranks. Golbal ranks make sense since will make most address checking a direct table look up -- more straight code and more efficient too.

am_send_hdr api is more to differentiate when upper layer don't have a request object, rather than the missing payload. The current api forces bad code factor as the former need tracking data at transport or inject emulation, and the latter affects lmt or rndz.

Unifying the api allows more sensible code factoring.

Reply:
One thing to note about the _reply APIs is that they were intended to be callback safe. I.e. they do not attempt to make progress. This is to ensure message handlers are called in order. On the ucx side there is no issue. It looks like ofi does call progress from reply functions. I think this is reason for the "unordered message" queue it maintains 😞

Reply:
Yes, I was wondering why we need this "unordered message" in OFI. I am still not sure I understand how that could happen. Could you make an artificial example how the callbacks may end-up out-of-order?

Reply:
In ch4/ofi progress, we read some number of elements from the completion queue and process them (see MPIDI_OFI_NUM_CQ_ENTRIES, default 8). If while processing those CQ elements we call a reply API, then it is possible we will again call into progress before we are all CQ elements have been processed. For instance we have to retry a reply message due to `FI_EAGAIN`, as you saw in #4355. The workaround added was to use an AM sequence number and defer processing of incoming messages if there is gap.

Calling progress from reply APIs was not supposed to be allowed be design, but this is the reality. I think to properly enforce the design, we would have to add an AM send queue into ch4/ofi and ensure it is flushed before sending new messages. Not sure if that is better or worse than the current deferred message processing...

Reply:
Actually the am part is using macro `MPIDI_OFI_CALL_RETRY_AM` (https://github.com/pmodels/mpich/blob/c6e08ce892ba9fa7dea4b647d1d57cbd42091737/src/mpid/ch4/netmod/ofi/ofi_am_impl.h#L57-L76), which enqueues the cq entries rather than process them. I wonder maybe that has something to do with what you described.

But even with `MPIDI_OFI_CALL_RETRY`, I still don't see how the problem arise.
```
OFI_Progress
    --> foreach cq entry, MPIDI_OFI_dispatch_function
        --> am_recv_event
            --> target_callback
                   --> am_send_reply
                          --> OFI_Progress
```
The inner `OFI_Progress` will get to process some active messages before some older ones from the first `OFI_Progress`. But I don't see how that is a problem. AM packets carries their own context so they shouldn't have dependency on the transport orders, right? And the dependency within a protocol chain is semantically controlled, ie if the previous packet from a single protocol hasn't been processed, it shouldn't have sent reply yet, so there isn't a possibility to have later packet (within a protocol) appear before early packets, right?

I understand sometime there is ordering requirement beyond single am message, e.g. at MPI-layer, but that is taken care of at the generic layer (with postponing queues) anyway.

Reply:
Hmm, I might be mistaken. Actually after commenting I was thinking about the internal ch4/ofi active message protocol and how that might perturb the order set by the generic layer. In ch4/ucx, messages are received and consumed as they arrive in order. But in ofi, message payloads may come later (see #4313). I believe smaller messages can overtake longer ones in that case, which the generic layer is not able to handle.

Reply:
> I understand sometime there is ordering requirement beyond single am message, e.g. at MPI-layer, but that is taken care of at the generic layer (with postponing queues) anyway.

Do we have postponing queues in the generic layer? My understanding of the generic layer is that it assumes messages are handled in order by the receiver. We only added sequence numbers in ch4 because of the need for the ofi workaround. Otherwise they would not be necessary. ch3 does not rely on them.

Reply:
> Do we have postponing queues in the generic layer?

It's here: https://github.com/pmodels/mpich/blob/c6e08ce892ba9fa7dea4b647d1d57cbd42091737/src/mpid/ch4/src/ch4r_callbacks.c#L50-L71

Reply:
Hmm, so I guess we do have that in the generic layer 😄. #3133 added more reorder logic in ch4/ofi. I approved it, but it's been a while and might be worth a revisit.

Reply:
Interesting. Do you recall we had test failures before that patch?

Reply:
Not in our infrastructure. IIRC, the problem was not observable with the sockets provider. I believe Intel observed it using psm2.